### PR TITLE
os/bluestore: Add missing punch_hole call so that object offsets are …

### PR DESCRIPTION
**Pull request by agayev**
…updated.

Rearranged an HM-SMR related code in _do_write_small so that it reuses
previously calculated offsets and, most importantly, added the missing
onode->extent_map.punch_hole call that releases overwritten blobs and updates
the location of the object to point to the latest copy.  Without it the OSD
could not restart becase a stale copy of osd_superblock object was being read.

Also added a fix to ZonedAllocator that stops the assertion from triggering when
a free region that spans two zones and ends at arbitrary bounds is added.  This
fix is temporary -- as the updated doc/dev/zoned-storage.rst explains, a new
FreelistManager implementation that is aware of the zone states will be added
for a proper fix.

Signed-off-by: Abutalib Aghayev <agayev@cs.cmu.edu>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by athanatos** [comment]
@aclamk Would you be able to have a look?

**Comment by athanatos** [comment]
@ifed01 Any chance you could have a look?

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -13064,6 +13048,22 @@ void BlueStore::_do_write_small(
...
+  // _do_write_small should not trigger.  OSDs, however, write and update a tiny
+  // amount of metadata, such as OSD maps, to disk.  For those cases, we
+  // temporarily just pad them to min_alloc_size and write them to a new place
+  // on every update.
+  if (bdev->is_smr()) {
+    BlobRef b = c->new_blob();
+    uint64_t b_off = p2phase<uint64_t>(offset, alloc_len);
+    uint64_t b_off0 = b_off;
+    _pad_zeros(&bl, &b_off0, min_alloc_size);
+    o->extent_map.punch_hole(c, offset, length, &wctx->old_extents);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:13062]: you should call o->extent_map.fault_range(db, offset, length) prior to punching a hole to make sure proper subset of extent map is loaded.

**Comment by agayev** [comment]
> Suggest to avoid merging multiple independent fixes in a single commit? May be split into 3 different ones: rst modification, small_write's and allocator's fixes?

Thanks for the review!  Split it into three commits.

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (agayev) [reply src/os/bluestore/BlueStore.cc:13071]: Done.  Moved the block below the existing extent_map.fault_range() call.

**Comment by aclamk** [comment]
@agayev I am looking for method to run validation tests for SMR related changed. Do you have some way to simulate zoned behaviour on regular block devices?

**Comment by agayev** [comment]
> @agayev I am looking for method to run validation tests for SMR related changed. Do you have some way to simulate zoned behaviour on regular block devices?
> By simulate I mean a simulator device that will complain if zoned requirements were violated.

@aclamk  There is this: https://zonedstorage.io/projects/tcmu-runner/.

**Comment by tchaikov** [comment]
jenkins test dashboard backend

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2020-06-19_01:26:25-rados-wip-kefu2-testing-2020-06-19-0046-distro-basic-smithi/

- cbt failures should be addressed by #35683
- cephadm failures are caused by DoS of docker.io

**Comment by agayev** [comment]
> > @agayev I am looking for method to run validation tests for SMR related changed. Do you have some way to simulate zoned behaviour on regular block devices?
> > By simulate I mean a simulator device that will complain if zoned requirements were violated.
> 
> @aclamk There is this: https://zonedstorage.io/projects/tcmu-runner/.

@aclamk Forgot to tell, there is also this: https://zonedstorage.io/linux/nullblk/.  It is much more lightweight.  You just insert null_blk module passing it parameters, and /dev/nullb0 appears as a zoned block device.

**Pull request merged**